lib/srv/db/redis/client.go

```diff
@@ -0,0 +1,205 @@
+	}
+}
+
+// Process add supports for additional cluster commands.
+func (c *clusterClient) Process(ctx context.Context, inCmd redis.Cmder) error {
+	cmd, ok := inCmd.(*redis.Cmd)
+	if !ok {
+		return trace.BadParameter("failed to cast Redis command")
+	}
+
+	switch cmdName := strings.ToLower(cmd.Name()); cmdName {
+	case mulitCmd, execCmd, watchCmd:
+		// do not allow transaction commands as they always fail on exec.
+		return trace.NotImplemented("%s is not supported in the cluster mode", cmdName)
+	case dbsizeCmd:
+		// use go-redis dbsize implementation. It returns size of all keys in the whole cluster instead of
+		// just currently connected instance.
+		res := c.DBSize(ctx)
+		cmd.SetVal(res.Val())
+		cmd.SetErr(res.Err())
+
+		return nil
+	case scanCmd:
+		var resultsKeys []string
+		var mtx sync.Mutex
+
+		if err := c.ForEachMaster(ctx, func(ctx context.Context, client *redis.Client) error {
+			scanCmd := redis.NewScanCmd(ctx, client.Process, cmd.Args()...)
+			err := client.Process(ctx, scanCmd)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			keys, _, err := scanCmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			mtx.Lock()
+			resultsKeys = append(resultsKeys, keys...)
+			mtx.Unlock()
+
+			return nil
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+
+		cmd := redis.NewCmd(ctx, cmd.Args())
+		cmd.SetVal(resultsKeys)
+
+		return nil
+	case keysCmd:
+		var resultsKeys []string
+		var mtx sync.Mutex
+
+		if err := c.ForEachMaster(ctx, func(ctx context.Context, client *redis.Client) error {
+			scanCmd := redis.NewStringSliceCmd(ctx, cmd.Args()...)
+			err := client.Process(ctx, scanCmd)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			keys, err := scanCmd.Result()
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			mtx.Lock()
+			resultsKeys = append(resultsKeys, keys...)
+			mtx.Unlock()
+
+			return nil
+		}); err != nil {
+			return trace.Wrap(err)
+		}
+
+		cmd.SetVal(resultsKeys)
+
+		return nil
+	case mgetCmd:
+		if len(cmd.Args()) == 1 {
+			return trace.BadParameter("wrong number of arguments for 'mget' command")
+		}
+
+		var resultsKeys []interface{}
+		var mtx sync.Mutex
+
+		keys := cmd.Args()[1:]
+		for _, k := range keys {
+			result := c.Get(ctx, k.(string))
+			if result.Err() == redis.Nil {
+				mtx.Lock()
+				resultsKeys = append(resultsKeys, redis.Nil)
+				mtx.Unlock()
+				continue
+			}
+
+			if result.Err() != nil {
+				cmd.SetErr(result.Err())
+				return trace.Wrap(result.Err())
+			}
+
+			mtx.Lock()
+			resultsKeys = append(resultsKeys, result.Val())
+			mtx.Unlock()
+		}
+
+		cmd.SetVal(resultsKeys)
+
+		return nil
+	case flushallCmd, flushdbCmd:
+		if err := c.ForEachMaster(ctx, func(ctx context.Context, client *redis.Client) error {
+			singleCmd := redis.NewCmd(ctx, cmd.Args()...)
+			err := client.Process(ctx, singleCmd)
+			if err != nil {
+				return trace.Wrap(err)
+			}
+
+			cmd.SetVal(singleCmd.Val())
+			cmd.SetErr(singleCmd.Err())
```

Comment:
race condition `c.ForEachMaster` calls the `ForEachMaster` in for each master asynchronously in new goroutine.

Reply:
Added mutex and error check above.